pkg/testrig/testrig.go

```diff
@@ -0,0 +1,221 @@
+// for every service, it creates main.go (optionally with DB connection) and a dockerfile to containerize it
+// if the service contains DB tables, it creates sidecar postgres container and creates a schema there
+// finally, it creates docker-compose.yml at the point of all (likely your project's root)
+// that joins all containerized services
+func GenerateRig(templateFile string, outputDir string, modules []*sysl.Module) error {
+	serviceMap, err := readTemplate(templateFile)
+	if err != nil {
+		return err
+	}
+
+	applications := make(map[string]*sysl.Application)
+	for _, module := range modules {
+		for appName, app := range module.GetApps() {
+			applications[appName] = app
+		}
+	}
+	appsWhoNeedDb := make(map[string]bool)
+	for serviceName := range serviceMap {
+		appsWhoNeedDb[serviceName] = false
+		if applications[serviceName] != nil {
+			if appNeedsDb(applications[serviceName]) {
+				appsWhoNeedDb[serviceName] = true
+			}
+		}
+	}
+
+	composeServices := make(map[string]interface{})
+
+	for serviceName, serviceTmpl := range serviceMap {
+		err = os.MkdirAll(filepath.Join(outputDir, serviceName), os.ModePerm)
+		if err != nil {
+			return err
+		}
+
+		err = generateMain(serviceName, serviceTmpl, outputDir, appsWhoNeedDb[serviceName])
+		if err != nil {
+			return err
+		}
+
+		err = generateDockerfile(serviceName, serviceTmpl, outputDir)
+		if err != nil {
+			return err
+		}
+
+		block := generateAppService(serviceName, serviceTmpl, outputDir)
+		composeServices[serviceName] = block
+		if appsWhoNeedDb[serviceName] {
+			block := generateDbService(serviceName)
+			composeServices[serviceName+"_db"] = block
+		}
+	}
+
+	composeFile, err := os.Create("docker-compose.yml")
+	if err != nil {
+		return err
+	}
+	defer composeFile.Close()
+	return generateCompose(composeFile, composeServices)
+}
+
+func appNeedsDb(app *sysl.Application) bool {
+	patterns := app.GetAttrs()["patterns"]
+	if patterns == nil {
+		return false
+	}
+	return strings.Contains(patterns.String(), "DB")
+}
+
+func readTemplate(templateFileName string) (template.ServiceMap, error) {
+	templateFile, err := os.Open(templateFileName)
+	if err != nil {
+		return nil, err
+	}
+	defer templateFile.Close()
+
+	byteValue, err := ioutil.ReadAll(templateFile)
+	if err != nil {
+		return nil, err
+	}
+	var vars template.ServiceMap
+	err = json.Unmarshal(byteValue, &vars)
+	if err != nil {
+		return nil, err
+	}
+
+	return vars, nil
+}
+
+func generateAppService(serviceName string, serviceTmpl template.Service, outputDir string) map[string]interface{} {
+	build := make(map[string]interface{})
+	build["context"] = "."
+	build["dockerfile"] = filepath.Join(outputDir, serviceName, "Dockerfile")
+	block := make(map[string]interface{})
+	block["build"] = build
+	block["ports"] = []string{fmt.Sprintf("%v:%v", serviceTmpl.Port, serviceTmpl.Port)}
+	return block
+}
+
+func generateDbService(serviceName string) map[string]interface{} {
+	block := make(map[string]interface{})
```

Comment:
Do this as a literal map instead of constructing it one element at a time:
```
	return map[string]interface{}{
		"image": "postgres:latest",
		"ports": []string{"5432:5432"},
		"volumes": []string{
			fmt.Sprintf("../gen/%v/%v.sql:/docker-entrypoint-initdb.d/%v.sql", serviceName, serviceName, serviceName),
		},
		"environment": map[string]string{
			"POSTGRES_USER":     "someuser",
			"POSTGRES_PASSWORD": "somepassword",
			"POSTGRES_DB":       "somedb",
		},
	}
```

Reply:
oh, brilliant, thank!